components/mediation/data-publishers/org.wso2.micro.integrator.analytics.messageflow.data.publisher/src/main/java/org/wso2/micro/integrator/analytics/messageflow/data/publisher/producer/AnalyticsDataProviderHolder.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.micro.integrator.analytics.messageflow.data.publisher.producer;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.config.SynapsePropertiesLoader;
+import org.wso2.micro.integrator.analytics.messageflow.data.publisher.publish.elasticsearch.ElasticConstants;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+
+/**
+ * Holder class to initialize and hold the custom data provider
+ */
+public class AnalyticsDataProviderHolder {
+    private static final Log log = LogFactory.getLog(AnalyticsDataProviderHolder.class);
+
+    private static AnalyticsCustomDataProvider analyticsCustomDataProvider = null;
+    private static volatile boolean isInitiationAttempted = false;
+
+    private AnalyticsDataProviderHolder() {
+    }
+
+    private void loadCustomDataProvider() {
+        try {
+            String analyticsCustomDataProviderClass = SynapsePropertiesLoader.getPropertyValue(
+                    ElasticConstants.SynapseConfigKeys.ELASTICSEARCH_CUSTOM_DATA_PROVIDER_CLASS, null);
+            isInitiationAttempted = true;
+            if (analyticsCustomDataProviderClass == null) {
+                return;
+            }
+            Class<?> clazz = Class.forName(analyticsCustomDataProviderClass);
+            Constructor<?> constructor = clazz.getConstructors()[0];
+            analyticsCustomDataProvider = (AnalyticsCustomDataProvider) constructor.newInstance();
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException
+                 | InvocationTargetException e) {
+            log.error("Error in obtaining custom producer class", e);
+        }
+    }
+
+    /**
+     * Create if absent and return the custom data provider
+     *
+     * @return Custom data provider
+     */
+    public static AnalyticsCustomDataProvider getAnalyticsCustomDataProvider() {
+        if (!isInitiationAttempted) {
+            synchronized (AnalyticsDataProviderHolder.class) {
+                if (!isInitiationAttempted) {
+                    new AnalyticsDataProviderHolder().loadCustomDataProvider();
+                }
+            }
+        }
+        return analyticsCustomDataProvider;
```

Comment:
@gayaldassanayake are we trying to do a singleton behaviour with this `isInitiationAttempted` variable? If so, shall we write this class as a singleton class?

Reply:
Fixed with 752787c548633c52d0e863ea053f7fa94c7a8cd4